Visual updates for Send To Nos

dcadenas opened the pull request:
I have the menus ready for review. Some things I know are pending and need more feedback:

1) The mapping between the categories we show and the categories we have (nip-56 and nip-69). See [issue comment](https://github.com/planetary-social/nos/issues/1012#issuecomment-2064933177)
2) I need to know if we need a confirmation alert for `Send To Nos` and a decision on what copy to use. The video below shows that I used as placeholder.
3) The encryption code itself which lives in issue #1015 

Here's what this PR has so far:

https://github.com/planetary-social/nos/assets/18184/47de5a34-553f-4f8d-aac9-b3b134d815d4

#1011 #1012 #1013 #1014

setch-l commented:
@dcadenas - Thanks for adding the confirmation for Submit to Nos. I would like to update the text as follows:
Confirm Flag 
Nos will analyze the note for Nudity or Sexual content and if we agree will label the note anonymously.

dcadenas commented:
One details here is that the bot will first analyze the text as it always does, and in this stage the category selected may not match the one proposed by the user. We may also manually choose another type of categorization. So we could add some text to describe this situation? that the category is treated more as a hint than a must

setch-l commented:
That's helpful context. Let's try this: 

Confirm Flag
Nos will analyze the note for Nudity or Sexual content and if we agree will label the note anonymously. If the label doesn't match the NIP-56 for the category we may choose a different label or no label at all.

dcadenas commented:
I noticed that the general note deletion alert had the title "Confirm Report" so I changed it to "Confirm Delete". Not directly related to the current issue but it's not worth to add a new issue just for that fix.

mplorentz commented on Nos/Views/ReportMenu.swift:
```diff
@@ -116,30 +122,90 @@ struct ReportMenuModifier: ViewModifier {
+    typealias TopLevelDisplayName = String
+    /// An enum to simplify the user selection through the sequence of connected
+    /// dialogs
+    enum UserSelection: Equatable {
```
I think we will need to split this eventually because we have one menu for reporting notes and another for reporting users. Maybe UserSelection could turn into a `protocol` like `ReportMenuItems` and then could have two implementations like `ReportContentMenuItems`and `ReportAuthorMenuItems` or similar. Or we could somehow pass the ReportTarget in each case and branch on it... that seems messier though.

dcadenas commented on Nos/Views/ReportMenu.swift:
```diff
@@ -116,30 +122,90 @@ struct ReportMenuModifier: ViewModifier {
+    /// List of the top-level report categories we care about
+    /// Out vocabulary is much bigger
+    func topLevelButtons() -> [ButtonState<UserSelection>] {
+        let spam = ("SP", "Spam")
```
@mplorentz I refactored this to be able to integrate the new display names to the categories data.
I also took the freedom to update some mappings because I think they were wrong because the illegal category wasn't used but we were using the illegal text. Also the only category we have that maps to nip-56 profanity was "coarse language" but we wanted to present that as "harrasment or intolerance" (harassment is already part of illegal subcategories). So I change the label to "coarse language or intolerance" which I think maps better to something that although it not illegal, could be reported. This also avoids breaking the mapping when we show reports from other clients that were used with the intent of coarse language but nothing to do with intolerance or harassment.

So I have this now, 
<img width="388" alt="Screenshot 2024-04-24 at 2 12 37 PM" src="https://github.com/planetary-social/nos/assets/18184/9e0051c5-1b49-4530-8d24-3b4716a0c388">



which in order, map to the following categories:
1. spam (unchanged, mapped to nip 56 spam),
2. nudity (changed the localized string which used to be 'Nudity & Sex', mapped to nip 56 nudity),
3. coarseLanguage (changed the label from 'Harassment or Intolerance', mapped to nip 56 profanity),
4. illegal (used to be violence which is mapped to nip 56 other, now mapped to nip 56 illegal),
5. other (unchanged, mapped to nip 56 other)

dcadenas replied:
@setch-l ^ I took the freedom to do some changes in the category mapping to avoid breaking the categories we have in cleanstr and we can reuse the names without clashes when displaying them from other sources

mplorentz replied:
These look good to me!

joshuatbrown commented on Nos/Models/ReportCategory.swift:
```diff
@@ -2,16 +2,16 @@ import Foundation
 /// A model for potential reasons why something might be reported.
 struct ReportCategory: Identifiable, Equatable {
-
+    
```
I wonder if Xcode is adding these trailing spaces; check Xcode > Settings > Text Editing > Editing and see if "Automatically trim trailing whitespace" is on as well as the sub-option for "Including whitespace-only lines".
<img width="942" alt="Screenshot 2024-04-25 at 8 59 17 AM" src="https://github.com/planetary-social/nos/assets/59564/4f746396-99af-41ba-a6e7-fe64830b613a">

dcadenas replied:
It's not turned on. I think it was one of my switches to VSCode. I added a `"files.trimTrailingWhitespace": true,` there

dcadenas commented on Nos/Models/ReportCategory.swift:
```diff
@@ -20,90 +20,102 @@ struct ReportCategory: Identifiable, Equatable {
-        ]
-    ),
-    ReportCategory(name: .moderation.other, code: "NA", nip56Code: .other),
+let coarseLanguage = ReportCategory(name: .moderation.coarseLanguage, code: "CL", nip56Code: .profanity)
```
Implemented in https://github.com/planetary-social/nos/pull/1063/commits/248cbc4584db35c8f9b647d3673ef7f83c1b9949

mplorentz commented:
@dcadenas I think you can mark this one ready for review instead of a draft?

mplorentz reviewed (commented):
This is coming along nicely. Was there anything in particular you wanted a review on at this stage? I think the main remaining work is:
1. Wire up the "Send to Nos" flow.
2. Separate the flow for reporting a note from the one for reporting an author, since they have different categories (and maybe messages?)

dcadenas commented:
@mplorentz I added the pending bifurcation for the different report user category list and integrated the rest of the PRs. 
I changed the code for the RelayService. I noticed that there's a way to wait for the ACK message from the server and only then I disconnect. This seemed to fix an issue I was having in which although the message was written, we were closing everything too early. We could specifically wait for an OK message from the server but I think any text message makes us more tolerant.

mplorentz commented on Nos/Models/ReportCategory.swift:
```diff
@@ -50,19 +51,19 @@ enum ReportCategoryType {
         nip56Code: .profanity
     )
-    
+
```
These changes are kind of annoying. Is this a setting from VSCode you could change?

mplorentz commented on Nos/Models/ReportCategory.swift:
```diff
@@ -245,7 +244,9 @@ let categories = [
 ]
 
-let selectableCategories = [
+let authorCategories = allCategories
```
I'm looking at the description for [this ticket](https://github.com/orgs/planetary-social/projects/11/views/1?filterQuery=moderation&pane=issue&itemId=60848564) and the categories listed there are the same as `noteCategories` here. I guess that is a separate ticket and assigned to @joshuatbrown, just pointing it out I guess that I think that ticket could be a one line change if branched off this PR.

mplorentz commented on Nos/Assets/Localization/Localizable.xcstrings:
```diff
@@ -10096,6 +10129,17 @@
+        "en" : {
+          "stringUnit" : {
+            "state" : "translated",
+            "value" : "Nos will analyze the note for %@ content and if we agree will label the note anonymously. If the label doesn't match the NIP-56 for the category we may choose a different label or no label at all."
```
I find this confusing on two points:
1. If Nos labels the note anonymously then how will people know to trust the events? I think what we are trying to say is that **you**, the user, will remain anonymous, but Nos will publish it from our Reportinator account.
2. "If the label doesn't match the NIP-56 for the category" - this doesn't make sense to me, and I don't think users will know what the difference between "label" and "category" mean to or what NIP-56 is.  

Here's a take at solving both issues:
"The Nos moderation team will analyze the note for %@ content and may publish a report from our own account, concealing your identity."

mplorentz replied:
Going to loop in @setch-l for proposed copy changes.

dcadenas replied:
Added 
<img width="426" alt="Screenshot 2024-04-29 at 3 49 34 PM" src="https://github.com/planetary-social/nos/assets/18184/c024c6d8-9af1-4fc3-b83e-955b399c23d2">

mplorentz commented on Nos/Views/ReportMenu.swift:
```diff
@@ -138,70 +150,62 @@ struct ReportMenuModifier: ViewModifier {
                 return String(localized: .localizable.reportSendToNosConfirmation(category.displayName))
             case .flagPublicly(let category):
                 return String(localized: .localizable.reportFlagPubliclyConfirmation(category.displayName))
```
I think we need to change this string to say "flag publicly" instead of "report publicly". I know that isn't necessarily specified in your ticket requirements but you already changed the title of this alert so we should change the description to match.

dcadenas replied:
Ok if I look for all "report" references in the localization file and change them to flag accordingly? or should we leave that for another ticket?

mplorentz replied:
That's should be fine. Although Josh changed one of them in another PR, so it might create conflicts.

dcadenas commented on Nos/Views/ReportMenu.swift:
```diff
@@ -75,41 +61,57 @@ struct ReportMenuModifier: ViewModifier {
+                    message: TextState(String(localized: .localizable.reportActionTitle(category.displayName))),
+                    buttons: [
+                        ButtonState(action: .send(.sendToNos(category))) {
+                            TextState("Send to Nos")
```
This is weird, I don't see it red. What could be wrong, needs a rebuild?

dcadenas replied:
<img width="432" alt="Screenshot 2024-04-29 at 2 15 42 PM" src="https://github.com/planetary-social/nos/assets/18184/4a9cc1a8-df52-4b6c-9b61-9dac3037c500">

mplorentz replied:
Ok I rebuilt and I don't see it now either 🤷‍♂️ I guess we are good.

setch-l commented:
@dcadenas - I'm fine with the proposed copy changes.

dcadenas commented:
@mplorentz I think you were referring to the copy @setch-l  now approved and I just added. I will move this then to UAT. Merge is done only after UAT right?

setch-l commented:
This looks good to me. We may get snagged on a few in the future but I think we've caught most of them.

joshuatbrown commented:
@dcadenas merge first, then it'll automatically appear in a Nos Staging for @setch-l to do UAT.
